### PR TITLE
Add support for long double fields

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature:
Unstructured index now default use long, double to replace int, float
Structured index now support create fields with type `long`, `double`, `array<long>`, `array<double>`

* **What is the current behavior?** (You can also link to an open issue here)
We use int, float for unstructured index
we only support int, float, array<int>, array<float> in structured index

* **What is the new behavior (if this is a feature change)?**
Unstructured index now default use long, double to replace int, float
Structured index now support create fields with type `long`, `double`, `array<long>`, `array<double>`

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no
no
* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
no

* **Related Python client changes** (link commit/PR here)
no

* **Related documentation changes** (link commit/PR here)
no

* **Other information**:
no

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: tests/tensor_search/integ_tests/test_add_documents_structured.py [hunk 1]
```diff
@@ -995,3 +1015,32 @@ def test_download_images_thread_count(self):
...
+            ({"long_field_1": 1}, False),  # small positive integer
+            ({"long_field_1": -1}, False),  # small negative integer
+            ({"long_field_1": 100232142864}, False),  # large positive integer that can't be handled by int
+            ({"long_field_1": -923217213}, False),  # large negative integer that can't be handled by int
+            ({'long_field_1': int("1" * 50)}, True),  # overlarge positive integer, should raise error in long field
+            # overlarge negative integer, should raise error in long field
+            ({'long_field_1': -1 * int("1" * 50)}, True),
+            ({"double_field_1": 1e10}, False),  # large positive integer mathematical expression
+            ({"double_field_1": -1e12}, False),  # large negative integer mathematical expression
+            ({"double_field_1": 1e10 + 0.123249357987123}, False),  # large positive float
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_add_documents_structured.py]: Are these the min/max acceptable values?

**Review by farshidz — CHANGES_REQUESTED** [review]
* Add input validation so out of range values don't cause 500. Input validation must take index version into account for unstructured (i.e. older indexes don't support double/long so lower limits)
* Ensure values at limit boundaries are tested, and also test past boudaries we get 400

**Comment by wanliAlex** [comment]
> * Add input validation so out of range values don't cause 500. Input validation must take index version into account for unstructured (i.e. older indexes don't support double/long so lower limits)
> * Ensure values at limit boundaries are tested, and also test past boudaries we get 400

I will add test cases at the boundaries.

As for the index version, I will leave it to manual, integration tests.

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py [hunk 2]
```diff
@@ -500,6 +508,52 @@ def _verify_marqo_field_type(self, field_name: str, value: Any):
...
+        MAX_FLOAT = 3.4028235e38
+        MIN_FLOAT = -3.4028235e38
+        SMALLEST_POS_FLOAT = 1.4e-45
+        LARGEST_NEG_FLOAT = -1.4e-45
+        if (value > MAX_FLOAT) or (0 < value < SMALLEST_POS_FLOAT) or (value < MIN_FLOAT) or \
+                (0 > value > LARGEST_NEG_FLOAT):
+            raise InvalidDataRangeError(f'Invalid value {value} for float field. Expected a value in the range '
+                                        f'[{MIN_FLOAT}, {MAX_FLOAT}] or [{LARGEST_NEG_FLOAT}, {SMALLEST_POS_FLOAT}], but '
+                                        f'found {value} '
+                                        f'If you wish to store a value outside of this range, creating a field with type '
```
Comment (farshidz) [inline src/marqo/core/structured_vespa_index/structured_vespa_index.py]: "create a field with type"

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py [hunk 3]
```diff
@@ -500,6 +508,52 @@ def _verify_marqo_field_type(self, field_name: str, value: Any):
...
+            self._verify_long_field_range(value)
+        elif index_field.type in (FieldType.Double, FieldType.ArrayDouble):
+            pass
+        else:
+            raise InternalError(f'Invalid field type {index_field.type} for field {index_field.name} called by'
+                                f'_verify_numerical_field_value. Expected one of {FieldType.Float}, {FieldType.Int}, '
+                                f'{FieldType.Long}, {FieldType.Double}.')
+
+    def _verify_float_field_range(self, value: float):
+        MAX_FLOAT = 3.4028235e38
```
Comment (farshidz) [inline src/marqo/core/structured_vespa_index/structured_vespa_index.py]: Best to define these at class level similar to `_VESPA_DOC_ID` so that they're stored only once not per function call, and also for clarity

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py [hunk 4]
```diff
@@ -500,6 +508,52 @@ def _verify_marqo_field_type(self, field_name: str, value: Any):
...
+        MIN_FLOAT = -3.4028235e38
+        SMALLEST_POS_FLOAT = 1.4e-45
+        LARGEST_NEG_FLOAT = -1.4e-45
+        if (value > MAX_FLOAT) or (0 < value < SMALLEST_POS_FLOAT) or (value < MIN_FLOAT) or \
```
Comment (farshidz) [inline src/marqo/core/structured_vespa_index/structured_vespa_index.py]: What happens if `< SMALLEST_POS_FLOAT` or `> LARGEST_NEG_FLOAT`? If truncation/rounding, I think that is acceptable and we shouldn't error out for that

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py [hunk 5]
```diff
@@ -500,6 +508,52 @@ def _verify_marqo_field_type(self, field_name: str, value: Any):
...
+            raise InvalidDataRangeError(f"Invalid value {value} for int field. Expected a value in the range "
+                                        f"[{MIN_INT}, {MAX_INT}], but found {value}. "
+                                        f"If you wish to store a value outside of this range, creating a field with type "
+                                        f"'{FieldType.Long} or '{FieldType.Double}'. ")
+
+    def _verify_long_field_range(self, value: int):
+        MAX_LONG = 9223372036854775807
+        MIN_LONG = -9223372036854775808
+        if value > MAX_LONG or value < MIN_LONG:
+            raise InvalidDataRangeError(f"Invalid value {value} for long field. Expected a value in the range "
```
Comment (farshidz) [inline src/marqo/core/structured_vespa_index/structured_vespa_index.py:557]: Does this actually work? Is Vespa's long range smaller than Python's?

File: tests/tensor_search/test_api.py [hunk 6]
```diff
@@ -115,11 +116,8 @@ def test_invalid_data_type(self):
                 }
             ]
         })
-
-        self.assertEqual(response.status_code, 400)
-        self.assertEqual(response.json()["code"], "invalid_argument")
-        self.assertEqual(response.json()["type"], "invalid_request")
-        assert "Expected a value of" in response.json()["message"] and "but found" in response.json()["message"]
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json()["errors"], True)
```
Comment (farshidz) [inline tests/tensor_search/test_api.py:120]: Why remove the message validation?

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py [hunk 7]
```diff
@@ -62,6 +63,13 @@ def to_vespa_document(self, marqo_document: Dict[str, Any]) -> Dict[str, Any]:
 
             index_field = self._marqo_index.field_map[marqo_field]
 
+            if isinstance(marqo_value, (int, float)) and not index_field.type == FieldType.Bool:
```
Comment (farshidz) [inline src/marqo/core/structured_vespa_index/structured_vespa_index.py]: Shouldn't need `not index_field.type == FieldType.Bool` because boolean conversion to int is done further down in line 73

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py (hunk 2 above)
Comment (wanliAlex) [reply src/marqo/core/structured_vespa_index/structured_vespa_index.py]: done

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py (hunk 5 above)
Comment (wanliAlex) [reply src/marqo/core/structured_vespa_index/structured_vespa_index.py:557]: Yes, Vespa will error out for ` int("1" * 30)`. We have a test case for this.

File: tests/tensor_search/test_api.py (hunk 6 above)
Comment (wanliAlex) [reply tests/tensor_search/test_api.py:120]: Add back a different message validation

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py (hunk 3 above)
Comment (wanliAlex) [reply src/marqo/core/structured_vespa_index/structured_vespa_index.py]: Make sense. Done

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py (hunk 7 above)
Comment (wanliAlex) [reply src/marqo/core/structured_vespa_index/structured_vespa_index.py]: We need to exclude the `boolean` type here. I will use `if type(marqo_value) in (float, int)` to check instead of `isinstance`

File: src/marqo/core/structured_vespa_index/structured_vespa_index.py (hunk 4 above)
Comment (wanliAlex) [reply src/marqo/core/structured_vespa_index/structured_vespa_index.py]: They will be round to 0. We think this is a reasonable behavior and we skip the check for that.

**Review by farshidz — APPROVED** [review]
